gpflow/optimizers/scipy.py

```diff
@@ -97,20 +114,34 @@ def initial_parameters(cls, variables: Sequence[tf.Variable]) -> tf.Tensor:
 
     @classmethod
     def eval_func(
-        cls, closure: LossClosure, variables: Sequence[tf.Variable], compile: bool = True
+        cls,
+        closure: LossClosure,
+        variables: Sequence[tf.Variable],
+        compile: bool = True,
+        unconnected_gradients: str = tf.UnconnectedGradients.ZERO,
+        gradient_already_checked: bool = True,
     ) -> Callable[[np.ndarray], Tuple[np.ndarray, np.ndarray]]:
         def _tf_eval(x: tf.Tensor) -> Tuple[tf.Tensor, tf.Tensor]:
             values = cls.unpack_tensors(variables, x)
             cls.assign_tensors(variables, values)
 
-            loss, grads = _compute_loss_and_gradients(closure, variables)
+            loss, grads = _compute_loss_and_gradients(closure, variables, unconnected_gradients)
             return loss, cls.pack_tensors(grads)
 
         if compile:
             _tf_eval = tf.function(_tf_eval)
 
         def _eval(x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
             loss, grad = _tf_eval(tf.convert_to_tensor(x))
+
+            if not gradient_already_checked and any(
+                [tf.math.count_nonzero(gr) == 0 for gr in grad]
```

Comment:
I don't think this technically is correct? You could have a zero-gradient for perfectly legitimate reasons? Also, we probably need to handle the case where the user sets `unconnected_gradients = tf.UnconnectedGradients.NONE`?